Flaky test_geo*

Some tests that match `test_geo*` in `tests/test_commands.py` and `tests/test_asyncio/test_commands.py` are not reliable.
We never observed them failing in CI or locally, but I received reports that they failed in some occasions.

These tests expect exact matches between floats e.g.

https://github.com/valkey-io/valkey-py/blob/6c95904b07bee87f6a6fd736a8557f166e48de53/tests/test_commands.py#L3544-L3563

And this doesn't always work.

According to the reports I received, the values are usually off by ~1e-15

Floats should be compared using [math.isclose](https://docs.python.org/3/library/math.html#math.isclose). The default tolerance should be enough.

These are the tests that have been reported as flaky:
```
FAILED tests/test_commands.py::TestValkeyCommands::test_geopos - assert [(2.1...
FAILED tests/test_commands.py::TestValkeyCommands::test_geosearch_member - As...
FAILED tests/test_commands.py::TestValkeyCommands::test_geosearch_with - Asse...
FAILED tests/test_commands.py::TestValkeyCommands::test_geosearchstore_dist
FAILED tests/test_commands.py::TestValkeyCommands::test_georadius_with - Asse...
FAILED tests/test_commands.py::TestValkeyCommands::test_georadius_store_dist
FAILED tests/test_commands.py::TestValkeyCommands::test_georadiusmember - Ass...
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_geopos[single]
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_geopos[pool]
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_georadius_with[single]
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_georadius_with[pool]
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_georadius_store_dist[single]
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_georadius_store_dist[pool]
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_georadiusmember[single]
FAILED tests/test_asyncio/test_commands.py::TestValkeyCommands::test_georadiusmember[pool]
```

here is the full build: https://koji.fedoraproject.org/koji/taskinfo?taskID=120194663

The error happens on s390x, it may be impossible to reproduce on other architectures.

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,4 +1,5 @@
 import argparse
+import math
 import time
 from typing import Callable, TypeVar
 from unittest import mock
@@ -491,6 +492,28 @@ def assert_resp_response(r, response, resp2_expected, resp3_expected):
         assert response == resp2_expected
     else:
         assert response == resp3_expected
+
+
+def assert_geo_is_close(coords, expected_coords):
+    """
+    Verifies that the coordinates are close within the floating point tolerance
+
+    Valkey uses 52-bit presicion
+    """
+    for a, b in zip(coords, expected_coords):
+        assert math.isclose(a, b)
+
+
+def assert_resp_response_isclose(r, response, resp2_expected, resp3_expected):
+    """Verifies that the responses are close within the floating point tolerance"""
+    protocol = get_protocol_version(r)
+
+    if protocol in [2, "2", None]:
+        assert_geo_is_close(response[0], resp2_expected[0])
+        assert response[1:] == resp2_expected[1:]
+    else:
+        assert_geo_is_close(response[0], resp3_expected[0])
+        assert response[1:] == resp3_expected[1:]
 
 
 def assert_resp_response_in(r, response, resp2_expected, resp3_expected):

--- a/tests/test_asyncio/test_commands.py
+++ b/tests/test_asyncio/test_commands.py
@@ -5,6 +5,7 @@ Tests async overrides of commands from their mixins
 import asyncio
 import binascii
 import datetime
+import math
 import re
 import sys
 from string import ascii_letters
@@ -13,8 +14,10 @@ import pytest
 import pytest_asyncio
 import valkey
 from tests.conftest import (
+    assert_geo_is_close,
     assert_resp_response,
     assert_resp_response_in,
+    assert_resp_response_isclose,
     is_resp2_connection,
     skip_if_server_version_gte,
     skip_if_server_version_lt,
@@ -2459,7 +2462,7 @@ class TestValkeyCommands:
 
         await r.geoadd("barcelona", values)
         # valkey uses 52 bits precision, hereby small errors may be introduced.
-        assert_resp_response(
+        assert_resp_response_isclose(
             r,
             await r.geopos("barcelona", "place1", "place2"),
             [
@@ -2519,7 +2522,30 @@ class TestValkeyCommands:
 
     @skip_unless_arch_bits(64)
     @skip_if_server_version_lt("3.2.0")
-    async def test_georadius_with(self, r: valkey.Valkey):
+    @pytest.mark.parametrize(
+        "georadius_kwargs, expected_georadius_result",
+        [
+            (
+                {"withdist": True, "withcoord": True, "withhash": True},
+                [b"place1", 0.0881, 3471609698139488],
+            ),
+            (
+                {"withdist": True, "withcoord": True},
+                [b"place1", 0.0881],
+            ),
+            (
+                {"withhash": True, "withcoord": True},
+                [b"place1", 3471609698139488],
+            ),
+            (
+                {"withdist": True, "withhash": True},
+                [b"place1", 0.0881, 3471609698139488],
+            ),
+        ],
+    )
+    async def test_georadius_with(
+        self, r: valkey.Valkey, georadius_kwargs, expected_georadius_result
+    ):
         values = (2.1909389952632, 41.433791470673, "place1") + (
             2.1873744593677,
             41.406342043777,
@@ -2530,33 +2556,23 @@ class TestValkeyCommands:
 
         # test a bunch of combinations to test the parse response
         # function.
-        assert await r.georadius(
+        georadius_result = await r.georadius(
             "barcelona",
             2.191,
             41.433,
             1,
             unit="km",
-            withdist=True,
-            withcoord=True,
-            withhash=True,
-        ) == [
-            [
-                b"place1",
-                0.0881,
-                3471609698139488,
-                (2.19093829393386841, 41.43379028184083523),
-            ]
-        ]
+            **georadius_kwargs,
+        )
 
-        assert await r.georadius(
-            "barcelona", 2.191, 41.433, 1, unit="km", withdist=True, withcoord=True
-        ) == [[b"place1", 0.0881, (2.19093829393386841, 41.43379028184083523)]]
-
-        assert await r.georadius(
-            "barcelona", 2.191, 41.433, 1, unit="km", withhash=True, withcoord=True
-        ) == [
-            [b"place1", 3471609698139488, (2.19093829393386841, 41.43379028184083523)]
-        ]
+        assert len(georadius_result) == 1
+        if "withcoord" in georadius_kwargs:
+            assert_geo_is_close(
+                georadius_result[0][-1], (2.19093829393386841, 41.43379028184083523)
+            )
+            assert georadius_result[0][:-1] == expected_georadius_result
+        else:
+            assert georadius_result == [expected_georadius_result]
 
         # test no values.
         assert (
@@ -2566,9 +2582,7 @@ class TestValkeyCommands:
                 1,
                 1,
                 unit="km",
-                withdist=True,
-                withcoord=True,
-                withhash=True,
+                **georadius_kwargs,
             )
             == []
         )
@@ -2632,7 +2646,8 @@ class TestValkeyCommands:
             "barcelona", 2.191, 41.433, 1000, store_dist="places_barcelona"
         )
         # instead of save the geo score, the distance is saved.
-        assert await r.zscore("places_barcelona", "place1") == 88.05060698409301
+        z_score = await r.zscore("places_barcelona", "place1")
+        assert math.isclose(z_score, 88.05060698409301)
 
     @skip_unless_arch_bits(64)
     @skip_if_server_version_lt("3.2.0")
@@ -2650,21 +2665,22 @@ class TestValkeyCommands:
         ]
         assert await r.georadiusbymember("barcelona", "place1", 10) == [b"place1"]
 
-        assert await r.georadiusbymember(
+        radius_place2, radius_place1 = await r.georadiusbymember(
             "barcelona", "place1", 4000, withdist=True, withcoord=True, withhash=True
-        ) == [
-            [
-                b"\x80place2",
-                3067.4157,
-                3471609625421029,
-                (2.187376320362091, 41.40634178640635),
-            ],
-            [
-                b"place1",
-                0.0,
-                3471609698139488,
-                (2.1909382939338684, 41.433790281840835),
-            ],
+        )
+
+        assert_geo_is_close(radius_place2[-1], (2.187376320362091, 41.40634178640635))
+        assert radius_place2[:-1] == [
+            b"\x80place2",
+            3067.4157,
+            3471609625421029,
+        ]
+
+        assert_geo_is_close(radius_place1[-1], (2.1909382939338684, 41.433790281840835))
+        assert radius_place1[:-1] == [
+            b"place1",
+            0.0,
+            3471609698139488,
         ]
 
     @skip_if_server_version_lt("5.0.0")

--- a/tests/test_commands.py
+++ b/tests/test_commands.py
@@ -1,5 +1,6 @@
 import binascii
 import datetime
+import math
 import re
 import threading
 import time
@@ -21,8 +22,10 @@ from valkey.client import EMPTY_RESPONSE, NEVER_DECODE
 
 from .conftest import (
     _get_client,
+    assert_geo_is_close,
     assert_resp_response,
     assert_resp_response_in,
+    assert_resp_response_isclose,
     is_resp2_connection,
     skip_if_server_version_gte,
     skip_if_server_version_lt,
@@ -3543,7 +3546,7 @@ class TestValkeyCommands:
         )
         r.geoadd("barcelona", values)
         # valkey uses 52 bits precision, hereby small errors may be introduced.
-        assert_resp_response(
+        assert_resp_response_isclose(
             r,
             r.geopos("barcelona", "place1", "place2"),
             [
@@ -3605,33 +3608,34 @@ class TestValkeyCommands:
         )
 
         r.geoadd("barcelona", values)
-        assert r.geosearch("barcelona", member="place1", radius=4000) == [
-            b"\x80place2",
-            b"place1",
-        ]
         assert r.geosearch("barcelona", member="place1", radius=10) == [b"place1"]
 
-        assert r.geosearch(
+        geosearch_place2, geosearch_place1 = r.geosearch(
             "barcelona",
             member="place1",
             radius=4000,
             withdist=True,
             withcoord=True,
             withhash=True,
-        ) == [
-            [
-                b"\x80place2",
-                3067.4157,
-                3471609625421029,
-                (2.187376320362091, 41.40634178640635),
-            ],
-            [
-                b"place1",
-                0.0,
-                3471609698139488,
-                (2.1909382939338684, 41.433790281840835),
-            ],
+        )
+
+        # All but the coordinates are identical
+        geosearch_place2[:-1] == [
+            b"\x80place2",
+            3067.4157,
+            3471609625421029,
         ]
+        assert_geo_is_close(
+            geosearch_place2[-1], (2.187376320362091, 41.40634178640635)
+        )
+        geosearch_place1[:-1] == [
+            b"place1",
+            0.0,
+            3471609698139488,
+        ]
+        assert_geo_is_close(
+            geosearch_place1[-1], (2.1909382939338684, 41.433790281840835)
+        )
 
     @skip_if_server_version_lt("6.2.0")
     def test_geosearch_sort(self, r):
@@ -3650,7 +3654,28 @@ class TestValkeyCommands:
 
     @skip_unless_arch_bits(64)
     @skip_if_server_version_lt("6.2.0")
-    def test_geosearch_with(self, r):
+    @pytest.mark.parametrize(
+        "geosearch_kwargs, expected_geosearch_result",
+        [
+            (
+                {"withdist": True, "withcoord": True, "withhash": True},
+                [b"place1", 0.0881, 3471609698139488],
+            ),
+            (
+                {"withdist": True, "withcoord": True},
+                [b"place1", 0.0881],
+            ),
+            (
+                {"withhash": True, "withcoord": True},
+                [b"place1", 3471609698139488],
+            ),
+            (
+                {"withdist": True, "withhash": True},
+                [b"place1", 0.0881, 3471609698139488],
+            ),
+        ],
+    )
+    def test_geosearch_with(self, r, geosearch_kwargs, expected_geosearch_result):
         values = (2.1909389952632, 41.433791470673, "place1") + (
             2.1873744593677,
             41.406342043777,
@@ -3660,44 +3685,23 @@ class TestValkeyCommands:
 
         # test a bunch of combinations to test the parse response
         # function.
-        assert r.geosearch(
+        geosearch_result = r.geosearch(
             "barcelona",
             longitude=2.191,
             latitude=41.433,
             radius=1,
             unit="km",
-            withdist=True,
-            withcoord=True,
-            withhash=True,
-        ) == [
-            [
-                b"place1",
-                0.0881,
-                3471609698139488,
-                (2.19093829393386841, 41.43379028184083523),
-            ]
-        ]
-        assert r.geosearch(
-            "barcelona",
-            longitude=2.191,
-            latitude=41.433,
-            radius=1,
-            unit="km",
-            withdist=True,
-            withcoord=True,
-        ) == [[b"place1", 0.0881, (2.19093829393386841, 41.43379028184083523)]]
-        assert r.geosearch(
-            "barcelona",
-            longitude=2.191,
-            latitude=41.433,
-            radius=1,
-            unit="km",
-            withhash=True,
-            withcoord=True,
-        ) == [
-            [b"place1", 3471609698139488, (2.19093829393386841, 41.43379028184083523)]
-        ]
-        # test no values.
+            **geosearch_kwargs,
+        )
+        assert len(geosearch_result) == 1
+        if "withcoord" in geosearch_kwargs:
+            assert_geo_is_close(
+                geosearch_result[0][-1], (2.1909382939338684, 41.433790281840835)
+            )
+            assert geosearch_result[0][:-1] == expected_geosearch_result
+        else:
+            assert geosearch_result == [expected_geosearch_result]
+
         assert (
             r.geosearch(
                 "barcelona",
@@ -3705,9 +3709,7 @@ class TestValkeyCommands:
                 latitude=1,
                 radius=1,
                 unit="km",
-                withdist=True,
-                withcoord=True,
-                withhash=True,
+                **geosearch_kwargs,
             )
             == []
         )
@@ -3793,7 +3795,7 @@ class TestValkeyCommands:
             storedist=True,
         )
         # instead of save the geo score, the distance is saved.
-        assert r.zscore("places_barcelona", "place1") == 88.05060698409301
+        assert math.isclose(r.zscore("places_barcelona", "place1"), 88.05060698409301)
 
     @skip_if_server_version_lt("3.2.0")
     def test_georadius_Issue2609(self, r):
@@ -3837,7 +3839,28 @@ class TestValkeyCommands:
 
     @skip_unless_arch_bits(64)
     @skip_if_server_version_lt("3.2.0")
-    def test_georadius_with(self, r):
+    @pytest.mark.parametrize(
+        "georadius_kwargs, expected_georadius_result",
+        [
+            (
+                {"withdist": True, "withcoord": True, "withhash": True},
+                [b"place1", 0.0881, 3471609698139488],
+            ),
+            (
+                {"withdist": True, "withcoord": True},
+                [b"place1", 0.0881],
+            ),
+            (
+                {"withhash": True, "withcoord": True},
+                [b"place1", 3471609698139488],
+            ),
+            (
+                {"withdist": True, "withhash": True},
+                [b"place1", 0.0881, 3471609698139488],
+            ),
+        ],
+    )
+    def test_georadius_with(self, r, georadius_kwargs, expected_georadius_result):
         values = (2.1909389952632, 41.433791470673, "place1") + (
             2.1873744593677,
             41.406342043777,
@@ -3848,48 +3871,26 @@ class TestValkeyCommands:
 
         # test a bunch of combinations to test the parse response
         # function.
-        assert r.georadius(
+        georadius_result = r.georadius(
             "barcelona",
             2.191,
             41.433,
             1,
             unit="km",
-            withdist=True,
-            withcoord=True,
-            withhash=True,
-        ) == [
-            [
-                b"place1",
-                0.0881,
-                3471609698139488,
-                (2.19093829393386841, 41.43379028184083523),
-            ]
-        ]
+            **georadius_kwargs,
+        )
 
-        assert r.georadius(
-            "barcelona", 2.191, 41.433, 1, unit="km", withdist=True, withcoord=True
-        ) == [[b"place1", 0.0881, (2.19093829393386841, 41.43379028184083523)]]
-
-        assert r.georadius(
-            "barcelona", 2.191, 41.433, 1, unit="km", withhash=True, withcoord=True
-        ) == [
-            [b"place1", 3471609698139488, (2.19093829393386841, 41.43379028184083523)]
-        ]
+        assert len(georadius_result) == 1
+        if "withcoord" in georadius_kwargs:
+            assert_geo_is_close(
+                georadius_result[0][-1], (2.19093829393386841, 41.43379028184083523)
+            )
+            assert georadius_result[0][:-1] == expected_georadius_result
+        else:
+            assert georadius_result == [expected_georadius_result]
 
         # test no values.
-        assert (
-            r.georadius(
-                "barcelona",
-                2,
-                1,
-                1,
-                unit="km",
-                withdist=True,
-                withcoord=True,
-                withhash=True,
-            )
-            == []
-        )
+        assert r.georadius("barcelona", 2, 1, 1, unit="km", **georadius_kwargs) == []
 
     @skip_if_server_version_lt("6.2.0")
     def test_georadius_count(self, r):
@@ -3949,11 +3950,14 @@ class TestValkeyCommands:
         r.geoadd("barcelona", values)
         r.georadius("barcelona", 2.191, 41.433, 1000, store_dist="places_barcelona")
         # instead of save the geo score, the distance is saved.
-        assert r.zscore("places_barcelona", "place1") == 88.05060698409301
+        assert math.isclose(r.zscore("places_barcelona", "place1"), 88.05060698409301)
 
     @skip_unless_arch_bits(64)
     @skip_if_server_version_lt("3.2.0")
     def test_georadiusmember(self, r):
+        """
+        Checks that r.georadiusbymember can return point, distance, and coordinate data
+        """
         values = (2.1909389952632, 41.433791470673, "place1") + (
             2.1873744593677,
             41.406342043777,
@@ -3961,28 +3965,32 @@ class TestValkeyCommands:
         )
 
         r.geoadd("barcelona", values)
-        assert r.georadiusbymember("barcelona", "place1", 4000) == [
+        assert r.georadiusbymember(name="barcelona", member="place1", radius=4000) == [
             b"\x80place2",
             b"place1",
         ]
         assert r.georadiusbymember("barcelona", "place1", 10) == [b"place1"]
 
-        assert r.georadiusbymember(
+        radius_place2, radius_place1 = r.georadiusbymember(
             "barcelona", "place1", 4000, withdist=True, withcoord=True, withhash=True
-        ) == [
-            [
-                b"\x80place2",
-                3067.4157,
-                3471609625421029,
-                (2.187376320362091, 41.40634178640635),
-            ],
-            [
-                b"place1",
-                0.0,
-                3471609698139488,
-                (2.1909382939338684, 41.433790281840835),
-            ],
+        )
+
+        # coordinates are in the last position
+
+        assert radius_place2[:-1] == [
+            b"\x80place2",
+            3067.4157,
+            3471609625421029,
         ]
+
+        assert_geo_is_close(radius_place2[-1], (2.187376320362091, 41.40634178640635))
+
+        assert radius_place1[:-1] == [
+            b"place1",
+            0.0,
+            3471609698139488,
+        ]
+        assert_geo_is_close(radius_place1[-1], (2.1909382939338684, 41.433790281840835))
 
     @skip_if_server_version_lt("6.2.0")
     def test_georadiusmember_count(self, r):

